### PR TITLE
msg/async/rdma: Zcopy preps

### PR DESCRIPTION
**Pull request by Adirl**

### CODE REVIEW

**Comment by Adirl** [comment]
@yuyuyu101 
pls see this preparations for replication zcopy
let me know what you think

**Comment by yuyuyu101** [comment]
I'm ok with these changes, but how to make zcopy happen?

**Comment by yuyuyu101** [comment]
Another important point:

ceph send each message with header, payload, data, footer. For all messages, we don't need to make header, payload, footer zerocopy. They are mutable when replicate to slave osds. Before I know the details about zcopy, I think you need to consider this and copy others except data to normal buffer. only make data registered memory zero copy.

**Comment by yuyuyu101** [comment]
If so, we could avoid most of memory hunger problem since in osd side, it may exist some unexpected behavior which hold bufferlist contains registered memory from rdma. These unexpected mostly related to non-data bufferlist.

File: src/msg/async/rdma/RDMAConnectedSocketImpl.cc [hunk 1]
```diff
@@ -454,7 +454,9 @@ ssize_t RDMAConnectedSocketImpl::submit(bool more)
   unsigned total = 0;
   unsigned need_reserve_bytes = 0;
   while (it != pending_bl.buffers().end()) {
-    if (infiniband->is_tx_buffer(it->raw_c_str())) {
+    if (infiniband->is_tx_buffer(it->raw_c_str()) && // TODO: tx zcopy
```
Comment (Adirl) [inline src/msg/async/rdma/RDMAConnectedSocketImpl.cc]: should be || and not &&

**Comment by Adirl** [comment]
@yuyuyu101 
should we parse incoming message for header, payload, data, footer and keep only the data buffer as reged memory ?

**Comment by Adirl** [comment]
@yuyuyu101 
sounds good, 
are you going to implement this?

**Comment by yuyuyu101** [comment]
yes, it would be todo action...

**Comment by Adirl** [comment]
@yuyuyu101 
here's an example for parsing incoming from xio code, 
src/msg/xio/XioConnection.cc:266:int XioConnection::handle_data_msg(struct xio_session *session,

i thought you might find it useful

**Comment by Adirl** [comment]
@DanielBar-On @saritz

**Comment by Adirl** [comment]
@yuyuyu101 
looking into AsyncConnection.cc::process()
maybe for "case STATE_OPEN_MESSAGE_READ_DATA" we can call read_until() with "keep_chunks" param 
makes sense? 

i set a meeting for two of us tomorrow to discuss

**Pull request closed without merging**